File: 2019/python/day1/main.py
from utils.timers import run_with_timer, get_data_with_timer


def get_data(filename):
	return [int(x.strip()) for x in open(filename).readlines()]


def part_one(d):
	return sum(calc_spec_fuel(i) for i in d)


def part_two(d):
	return sum(calc_total_fuel(i) for i in d)


def calc_spec_fuel(val):
	return (val // 3) - 2


def calc_total_fuel(start):
	n = calc_spec_fuel(start)
	return 0 if n <= 0 else n + calc_total_fuel(n)


def main(f="input.txt"):
	data = get_data_with_timer(get_data, f)
	run_with_timer(part_one, data)
	run_with_timer(part_two, data)


if __name__ == '__main__':
	main()

File: 2019/julia/src/day1/main.jl
data = map(x -> parse(Int64, x), readlines("input.txt"))

calc_spec_fuel(val) = convert(Int64, floor(val / 3)) - 2
part_one() = sum(map(calc_spec_fuel, data))
part_two() = sum(map(calc_total_fuel, data))

function calc_total_fuel(start)
	n = calc_spec_fuel(start)
	if n <= 0
		return 0
	end
	return n + calc_total_fuel(n)
end

function main()
   println(part_one())
   println(part_two())
end

main()
